app/controllers/courses_controller.rb

```diff
@@ -59,9 +59,21 @@ def update
 
   def destroy
     validate
-    DeleteCourseWorker.schedule_deletion(course: @course, current_user:)
-    render json: { success: true }
-  end
+    if params.key?(:campaign)
+      campaigns_course = CampaignsCourses.find_by(course_id: @course.id,
+                                                  campaign_id: params[:campaign_id])
+      result = campaigns_course&.destroy
+      message = result ? 'campaign.course_removed' : 'campaign.course_already_removed'
+      flash[:notice] = t(message, title: @course.title,
+                                  campaign_title: params[:campaign_title])
+    end
+    DeleteCourseWorker.schedule_deletion(course: @course, current_user: current_user)
+    if params.key?(:campaign)
+      redirect_to programs_campaign_path(params[:slug])
+    else
+      render json: { success: true }
+    end
+  end  
```

Comment:
This controller is already too large, and rather than make it even more complex, I think this is a good opportunity to move the destroy method to a child controller following the pattern of the other ones in `app/controllers/courses/`.